### PR TITLE
context: add basic context profiling

### PR DESCRIPTION
**Pull request by carl-mastrangelo**
This will allow us to measure how often contexts of particular sizes are used.

### CODE REVIEW

File: context/src/main/java/io/grpc/PersistentHashArrayMappedTrie.java [hunk 1]
```diff
@@ -221,13 +245,21 @@ public V get(K key, int hash, int bitsConsumed) {
             newValues,
             compressedIndex + 1,
             values.length - compressedIndex);
-        return new CompressedIndex<K,V>(newBitmap, newValues);
+        int newSize = size();
+        newSize += 1; // The new leaf
+        for (int i = compressedIndex; i < values.length; i++) {
```
Comment (zpencer) [inline context/src/main/java/io/grpc/PersistentHashArrayMappedTrie.java]: I think we just want to `size() + 1` here. The two arraycopy calls ensure all the original items are in the new array.

File: context/src/main/java/io/grpc/Context.java [hunk 2]
```diff
@@ -316,23 +365,49 @@ public CancellableContext withDeadline(Deadline deadline,
    *
    */
   public <V> Context withValue(Key<V> k1, V v1) {
-    return new Context(this, keyValueEntries.put(k1, v1));
+    PersistentHashArrayMappedTrie<Key<?>, Object> newKeyValueEntries = keyValueEntries.put(k1, v1);
+    if (shouldSample()) {
+      withValueDupeCounts.addAndGet(
```
Comment (zpencer) [inline context/src/main/java/io/grpc/Context.java]: It's not obvious what withValueDupeCounts is tracking, can we add a comment?

File: context/src/main/java/io/grpc/PersistentHashArrayMappedTrie.java (hunk 1 above)
Comment (carl-mastrangelo) [reply context/src/main/java/io/grpc/PersistentHashArrayMappedTrie.java]: Done.

File: context/src/main/java/io/grpc/Context.java (hunk 2 above)
Comment (carl-mastrangelo) [reply context/src/main/java/io/grpc/Context.java]: Done.

Comment (zpencer) [reply context/src/main/java/io/grpc/Context.java]: I think it should be `N - oldSize + newSize` for this stat to work, like if we added a key and the size increased from 3 to 4:

1 + 4 - 3 = 2 (right now we report 2 duplicates)
Instead if it's flipped we get:
1 + 3 - 4 = 0 (report 0 duplicates)

**Pull request merged**